 STEP 1 DATA Loading and Inspection

In [2]:
import pandas as pd
 
#df = pd.read_excel("F:/PYTHON PROJ/LEARN PYTHON/Customer Churn Prediction/Telco_customer_churn.xlsx", index_col=0)

df =pd.read_csv("F:\PYTHON PROJ\LEARN PYTHON\Customer Churn Prediction\WA_Fn-UseC_-Telco-Customer-Churn.csv")

df.head()


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


1.1 Basic Data Inspection

In [3]:
df.info()
df.shape
df.customerID.isnull().sum()
df.isnull().values.any()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

np.False_

Key Questions
1. There are  7043 entries, 21 cols
2. 18 str type and 2 int ('SeniorCitizen', 'tenure'), 1 float 'MonthlyCharges', 
3. df.info and df.isnull().values.any()does not work here since the pandas will treat " " not as NaN
    to check that 
    (df==" ").sum()


1.2 Target Variable (Churn)

1.2.1 check churn col  to  see the churned or unchurned group

In [4]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

convert qty to portion.

In [5]:
df['Churn'].value_counts(normalize=True)


Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

The dataset is moderately imbalanced, with significantly more non-churn customers than churn customers.
In later stages, techniques like resampling or class weighting can be used to handle imbalance.

1.3 Check Missing Values

In [6]:
(df == " ").sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [7]:
df.groupby('Churn').MonthlyCharges.mean()

Churn
No     61.265124
Yes    74.441332
Name: MonthlyCharges, dtype: float64

To clean the " ", we can first change it to NaN and then convert them to numeric
1.3.1 change it to NaN

In [8]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')


1.3.2. check total null

In [9]:
df.TotalCharges.isnull().sum()

np.int64(11)

1.3.3. Fill the missing value as median   (why median?) 1. reduce the outlier effect. 2. more stable than mean.

In [10]:
df.TotalCharges.fillna(df.TotalCharges.median())

0         29.85
1       1889.50
2        108.15
3       1840.75
4        151.65
         ...   
7038    1990.50
7039    7362.90
7040     346.45
7041     306.60
7042    6844.50
Name: TotalCharges, Length: 7043, dtype: float64

1.4 Numerical Feature Analysis

In [11]:
df['MonthlyCharges'].describe()

count    7043.000000
mean       64.761692
std        30.090047
min        18.250000
25%        35.500000
50%        70.350000
75%        89.850000
max       118.750000
Name: MonthlyCharges, dtype: float64

In [12]:
df.groupby('Churn')['MonthlyCharges'].mean()

Churn
No     61.265124
Yes    74.441332
Name: MonthlyCharges, dtype: float64

Insight:
Customers who churn tend to have higher monthly charges compared to those who stay.

This suggests that pricing may be a key factor contributing to churn, potentially due to perceived lower value or affordability issues.

In [13]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


1.5 categorical feature analysis

explore the possible relationship beween contract type and churn rates

In [14]:
df.groupby('Contract').Churn.value_counts(normalize=True)

Contract        Churn
Month-to-month  No       0.572903
                Yes      0.427097
One year        No       0.887305
                Yes      0.112695
Two year        No       0.971681
                Yes      0.028319
Name: proportion, dtype: float64

Contract Analysis:

- Month-to-month contracts have the highest churn rate
- Customers with longer contracts (one-year, two-year) have significantly lower churn rates

Insight:
There is a clear negative relationship between contract length and churn. Customers with longer-term commitments are less likely to churn.

1.5.2 More exploration

In [15]:
df.groupby('MultipleLines').Churn.value_counts(normalize=True)

MultipleLines     Churn
No                No       0.749558
                  Yes      0.250442
No phone service  No       0.750733
                  Yes      0.249267
Yes               No       0.713901
                  Yes      0.286099
Name: proportion, dtype: float64

Not significant churn rates between multiple lines

In [16]:
df.groupby('PhoneService').Churn.value_counts(normalize=True)

PhoneService  Churn
No            No       0.750733
              Yes      0.249267
Yes           No       0.732904
              Yes      0.267096
Name: proportion, dtype: float64

Not significant churn rates between phoneservice

In [17]:
df.groupby('InternetService').Churn.value_counts(normalize=True)

InternetService  Churn
DSL              No       0.810409
                 Yes      0.189591
Fiber optic      No       0.581072
                 Yes      0.418928
No               No       0.925950
                 Yes      0.074050
Name: proportion, dtype: float64

 
Internet Service Analysis:

- Customers using Fiber optic have the highest churn rate
- DSL users have moderate churn
- Customers with no internet service have the lowest churn rate

Insight:

Customers with higher-tier internet services (especially Fiber optic) are more likely to churn compared to those with basic or no internet service.

`It may also indicate that Fiber optic services are priced higher or face stronger competition, leading to increased churn risk.`

In [18]:
df.groupby('PaymentMethod').Churn.value_counts(normalize=True)

PaymentMethod              Churn
Bank transfer (automatic)  No       0.832902
                           Yes      0.167098
Credit card (automatic)    No       0.847569
                           Yes      0.152431
Electronic check           No       0.547146
                           Yes      0.452854
Mailed check               No       0.808933
                           Yes      0.191067
Name: proportion, dtype: float64

Payment Method Analysis:

- Customers using electronic check have the highest churn rate
- Other payment methods show relatively similar churn rates with no significant differences

Insight:

Payment method appears to have some impact on churn, particularly for electronic check users, while other methods do not show strong variation.

This may indicate that electronic check users are less engaged or less committed, or that this payment method is associated with a less stable customer segment.

It could also suggest that customers using automatic payment methods are more likely to stay due to convenience and lower friction.

STEP 1 SUMMARY

In my exploratory data analysis, I first examined the target variable and found that the churn rate is around 26.6%, indicating a moderate class imbalance.

I then analyzed numerical features and found that customers who churn tend to have higher monthly charges, suggesting that pricing may be a key factor driving churn.

Additionally, I explored categorical features and found that customers with month-to-month contracts have the highest churn rate, while those with longer-term contracts have significantly lower churn.

This indicates that contract commitment plays an important role in customer retention, and encouraging longer-term contracts could be an effective strategy to reduce churn.

The feature might be included in the ML 
1. 'Contract'
2. 'MonthlyCharges'
3. 'InternetService'
4. 'PaymentMethod'